core/src/proto/types.rs

```diff
@@ -38,416 +54,1077 @@ pub enum Error {
-/// non-malleable signature.
+impl From<BuilderSerde> for Vec<u8> {
+    fn from(tx: BuilderSerde) -> Vec<u8> {
+        tx.0
+    }
+}
+
+/// A structure to facilitate Serde (de)serializations of SaplingMetadata
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct SaplingMetadataSerde(Vec<u8>);
+
+impl From<Vec<u8>> for SaplingMetadataSerde {
+    fn from(tx: Vec<u8>) -> Self {
+        Self(tx)
+    }
+}
+
+impl From<SaplingMetadataSerde> for Vec<u8> {
+    fn from(tx: SaplingMetadataSerde) -> Vec<u8> {
+        tx.0
+    }
+}
+
+/// A section providing the auxiliary inputs used to construct a MASP
+/// transaction
 #[derive(
-    Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, BorshSchema, Hash,
+    Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
 )]
-pub struct SigningTx {
-    pub code_hash: [u8; 32],
-    pub data: Option<Vec<u8>>,
-    pub timestamp: DateTimeUtc,
+pub struct MaspBuilder {
+    /// The MASP transaction that this section witnesses
+    pub target: crate::types::hash::Hash,
+    /// The decoded set of asset types used by the transaction. Useful for
+    /// offline wallets trying to display AssetTypes.
+    pub asset_types: HashSet<(Address, Epoch)>,
+    /// Track how Info objects map to descriptors and outputs
+    #[serde(
+        serialize_with = "borsh_serde::<SaplingMetadataSerde, _>",
+        deserialize_with = "serde_borsh::<SaplingMetadataSerde, _, _>"
+    )]
+    pub metadata: SaplingMetadata,
+    /// The data that was used to construct the target transaction
+    #[serde(
+        serialize_with = "borsh_serde::<BuilderSerde, _>",
+        deserialize_with = "serde_borsh::<BuilderSerde, _, _>"
+    )]
+    pub builder: Builder<(), (), ExtendedFullViewingKey, ()>,
+}
+
+impl MaspBuilder {
+    /// Get the hash of this ciphertext section. This operation is done in such
+    /// a way it matches the hash of the type pun
+    pub fn hash<'a>(&self, hasher: &'a mut Sha256) -> &'a mut Sha256 {
+        hasher.update(
+            self.try_to_vec().expect("unable to serialize MASP builder"),
+        );
+        hasher
+    }
+}
+
+impl borsh::BorshSchema for MaspBuilder {
+    fn add_definitions_recursively(
+        _definitions: &mut std::collections::HashMap<
+            borsh::schema::Declaration,
+            borsh::schema::Definition,
+        >,
+    ) {
+    }
+
+    fn declaration() -> borsh::schema::Declaration {
+        "Builder".into()
+    }
+}
+
+/// A section of a transaction. Carries an independent piece of information
+/// necessary for the processing of a transaction.
+#[derive(
+    Clone,
+    Debug,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Serialize,
+    Deserialize,
+)]
+pub enum Section {
+    /// Transaction data that needs to be sent to hardware wallets
+    Data(Data),
+    /// Transaction data that does not need to be sent to hardware wallets
+    ExtraData(Code),
```

Comment:
Is this meant to be `ExtraData(Code)`?
```suggestion
    ExtraData(Data),
```

Reply:
Yes, it is meant to be `ExtraData(Code)`. The reason why is that the extra data section shares a greater resemblance to the the code section than it does to the data section. More specifically, extra data sections frequently contain too many bytes, just like the WASM code in code sections are frequently too large. Hence both extra data and code sections are "compressed" by first taking the SHA-256 hash of their payload and putting that into the section instead. This is in contrast to the data section, which is mostly unique to each transaction, and is never compressed because hardware wallets need to render all contained data (which would not be possible with just a hash).